_overviews/scala3-macros/tutorial/reflection.md

```diff
@@ -31,11 +41,126 @@ As it is a subtype of `Tree` we can also look into the `TreeMethods` to find mor
-<!-- Safty -->
-*Coming soon*
+## Relation with Expr/Type
+
+Expressions `Expr[T]` can be seen as wrapper around a `Term` that knows statically the type of the term.
+Bellow we use the `asTerm` extension method to transform the expression into a term.
+This extension method is only available after importing `quotes.reflect.asTerm`.
+Then we use `asExprOf[Int]` to transform the term back into `Expr[Int]`.
+This operation will fail if the term does not have provided type or if the term is not a valued expression.
+
+```scala
+def f(x: Expr[Int])(using Quotes): Expr[Int] =
+  import quotes.reflect.*
+  val tree: Term = x.asTerm
+  val expr: Expr[Int] = tree.asExprOf[Int]
+  expr
+```
+
+Similarly, we can also a `Type[T]` as a wrapper over a `TypeRepr` that knows statically its type.
+To get a `TypeRepr` we use `TypeRepr.of[X]` which expects a given `Type[X]` in scope (similar to `Type.of[X]`).
+We can also transform it back into a `Type[?]` using the `asType` method.
+As the type of `Type[?]` is not statically know we need to name it with an existential type to use it, can be done using the `'[t]` pattern.
+
+```scala
+def g[T: Type](using Quotes) =
+  import quotes.reflect.*
+  val tpe: TypeRepr = TypeRepr.of[T]
+  tpe.asType match
+    case '[t] => '{ val x: t = ${...} }
+  ...
+```
+
+## Macro API design
+
+It will be often useful to create helper methods or extractors that to perform some common logic of your macros.
+
+The simples methods will be those that only mention `Expr`, `Type` and `Quotes` in their signature.
+Internally they may use reflection but this will not be seen at the use site of the method.
+
+```scala
+def f(x: Expr[Int])(using Quotes): Expr[Int] =
+  import quotes.reflect.*
+  ...
+```
+
+In some cases it is inevitable to require some methods that work on `Tree`s or other types in `quotes.reflect`.
+For these cases, the best is to follow the following example of method signatures.
 
+A method that takes a `quotes.reflect.Term` parameter
+```scala
+def f(using Quotes)(term: quotes.reflect.Term): String =
+  import quotes.reflect.*
+  ...
+```
+
+An extension method for a `quotes.reflect.Term` returning a `quotes.reflect.Tree`
+```scala
+extension (using Quotes)(term: quotes.reflect.Term)
+  def g: quotes.reflect.Tree = ...
+```
+
+An extractor that matches on `quotes.reflect.Term`s.
+```scala
+object MyExtractor:
+  def unapply(using Quotes)(x: quotes.reflect.Term) =
+    ...
+    Some(y)
+```
+
+> **Avoid saving the `Quotes` context in a field.**
+> `Quotes` in fields inevitably make its use harder by hitting errors involving `Quotes` with different paths.
+>
+> Usually these patterns have been seen in code that uses the Scala 2 ways to define extension methods or contextual unapplies.
+> Now that we have `given` parameters that can be added before other parameters, all these old workarounds are no needed.
+> The new abstraction make it simpler on the definition site, use site and also have a better performance.
```

Comment:
```suggestion
> The new abstraction make it simpler on the definition site, use site, and also have a better performance.
```

Reply:
What are these peformance gains here?

Reply:
no object instantiations for extension methods.